Initial implementation

smarnach opened the pull request:
## ActiveTable XBlock

This XBlock implements a tabular problem type, where students are required to fill in some of the table cells with the correct answers.  The score awarded to the student is the proportion of correct answers.

The XBlock is implemented in a straight-forward way.  All data is stored in XBlock fields.  The content is definied by a table definition in a Python-like syntax and some supporting fields.  Student state consists of the answers students gave, the correctness of these answers and the score awarded.  The amount of data (both content and student state) is very small (likely to be far less than 1K per problem per student).  Since the amount of data is so small, and the XBlock doesn't do anything that is computationally intensive, we don't expect that there are any performance bottlenecks.  We haven't performed any load testing.

We'd like to get this XBlock on edx.org for Davidson College by mid-December.

For testing on the devstack, install the XBlock using

```
pip install -e .
```

as user edxapp from inside the repo directory and add `activetable` to the Advanced Module List in your course settings.  Most of the features can be tested directly with the example table you get when adding an ActiveTable component to a unit.

---

**Partner information** hosted on edx.org (Davidson College)
**Merge deadline** mid-December
**JIRA Story** [OSPR-923](https://openedx.atlassian.net/browse/OSPR-923)
**Confluence / Product Asset** N/A
**Sandbox URL** [LMS](http://pr10476.sandbox.opencraft.com/) ([Example](http://pr10476.sandbox.opencraft.com/courses/course-v1:OpenCraftX+1+2015/courseware/18d0e16c3b2842e5800abbadabf78ac0/77810af2829b4ff5acaee8f4002365bf/1)), [Studio](http://studio.pr10476.sandbox.opencraft.com/) ([Example](http://studio.pr10476.sandbox.opencraft.com/container/block-v1:OpenCraftX+1+2015+type@vertical+block@7d237ce535f6498cb51843db77f05a82))
**Dependencies** N/A

---

**PR Author(s) Notes / To-Do**
- [x] Fixed: Some integration tests are still missing.
- [x] Fixed: The images indicating correctness status do not work in Studio, probably because the static URL isn't properly rewritten.  The URL rewriting is some undocumented black magic which I didn't manage to debug so far.
- [x] Implemented the requested feature: I didn't reimplement all the features available in Capa problems, like setting a maximum attempt number or randomization.  Is there a way to reuse the Capa implementation instead of reimplementing this stuff?
- [x] Won't fix: Resetting the user state when logged in as a staff user in the LMS doesn not work for this XBlock.  I've got no idea how to fix this.

---

**Screenshots** TBD

---

**Reviewers**
- [x] Code: (@Kelketek)
- [ ] Code: (TBD)
- [ ] Product: (TBD)

smarnach committed:
Add skeleton created by the XBlock SDK.

smarnach committed:
First feature-complete version.

smarnach committed:
Add status message indicating the number of correct cells.

smarnach committed:
Apply a bunch of style fixes.

smarnach committed:
Add status image indicating answer correctness.

smarnach committed:
Add README.md file.

smarnach committed:
Accessibility improvements.

smarnach committed:
Configure pylint and make it happy.

smarnach committed:
Add unit tests for parsers.py.

smarnach committed:
Add display name for nicer title in Studio.

smarnach committed:
Add grading functionality.

smarnach committed:
Add unit tests for cells.py.

smarnach committed:
Add unit tests for activetable.py.

Kelketek commented on README.md:
```diff
@@ -0,0 +1,53 @@
+        ['French Revolution', Numeric(answer=1789)],
+        ['Krakatoa volcano explosion', Numeric(answer=1883)],
+        ["Proof of Fermat's last theorem", Numeric(answer=1994)],
+    ]
```
Now that we're using XBlocks, we can (and should) be designing these sorts of things with Child blocks if we can. This work of doing custom parsing is a lot to maintain when we have things like XBlock-utils to automate much of our form building. If we don't do this now, we're going to be stuck with it.

smarnach replied:
@Kelketek The client explicitly requested to not turn this into a point-and-click interface.  See Xavier's comments on the ticket for details.

I think child XBlocks would be a bad fit for this anyway.  Making every cell an XBlock would make creating tables really awkward, and it wouldn't be particularly efficient performance-wise either.  What I could imagine is to design a tailored graphical interface in JavaScript, but, as mentioned above, we were explicitly asked not to do that.

Kelketek replied:
Very well.

Kelketek commented on activetable/activetable.py:
```diff
@@ -0,0 +1,223 @@
+        scope=Scope.settings,
+        default='ActiveTable problem'
+    )
+    table_definition = String(
```
See the above comment about child blocks. I really think we should be using them here-- this methodology was built out of limitations that don't exist in XBlock.

Kelketek commented on activetable/static/README.txt:
```diff
@@ -0,0 +1,19 @@
+You can ask the runtime for a URL that will retrieve these files with:
+
+    url = self.runtime.local_resource_url(self, "static/js/lib.js")
+
```
You probably don't need this file-- it was generated by the SDK, yes?

smarnach replied:
You are right – I didn't even notice the file.  I'll remove it.

Kelketek commented:
@smarnach I'm worried about the approach of parsing the table manually now that we no longer need to.

As for your questions:

> I didn't reimplement all the features available in Capa problems, like setting a maximum attempt number or randomization. Is there a way to reuse the Capa implementation instead of reimplementing this stuff?

I don't think so. You will probably have to re-implement it. :/

> Resetting the user state when logged in as a staff user in the LMS doesn not work for this XBlock. I've got no idea how to fix this.

This is provided by the XModule shim. I don't think you'll be expected to fix it. You're talking about the button, right? There should still be the possibility of resetting it manually in the Instructor Dashboard, given the problem ID, as I think that's available for all blocks.

> The images indicating correctness status do not work in Studio, probably because the static URL isn't properly rewritten. The URL rewriting is some undocumented black magic which I didn't manage to debug so far.

Check how the problem-builder block handles this.

smarnach committed:
Remove redundant file generated by the SDK.

smarnach commented:
@Kelketek Thanks for your suggestions.

> I don't think so. You will probably have to re-implement it. :/

There are quite a few Capa features, so I asked Xavier which subset to preserve.

> This is provided by the XModule shim. I don't think you'll be expected to fix it. …

Ah, that makes sense.  And good to know there's a reset function in the Instructor Dashboard – that's very handy for testing.

> Check how the problem-builder block handles this.

Problem Builder doesn't seem to inlcude any static images.  After digging a bit further, I found that there is an API method to get resource URLs from the runtime, but [edx-platform doesn't support it](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/x_module.py#L1437-L1441).  I'm now copying the images into the XBlock, but it's still awkward to get the correct URLs in a CSS file.

smarnach committed:
Fix status icon URLs and include icons in the repository.

smarnach committed:
Remove __init__() method of ActiveTableXBlock.

According to the docstring of the XBlock class, subclasses are not allowed to
override __init__().  (Seriously?)

smarnach commented:
@Kelketek This is ready for another look now.  I'll wait for Xavier's answer before implementing any additional Capa features.

smarnach committed:
Make pep8 tool happy.

smarnach committed:
Configure Travis.

Kelketek commented on activetable/activetable.py:
```diff
@@ -0,0 +1,225 @@
+        'will result in equal-width columns with a total width of 800 pixels.',
+        scope=Scope.content,
+    )
+    row_heights = String(
```
The row heights and column widths fields have a reset button next to them in the interface. However, this button does not seem to work. Clicking it makes the field appear empty, but saving it does not actually save an empty string-- it just reverts to whatever it was before you hit the button. Putting in a new value does get saved, though.

Kelketek commented:
@smarnach This seems good to me so far. Find out from Davidson which of the CAPA features they need reimplemented and make the tests pass-- they're currently failing for a lack of the `yaml` module. You may wish to mimic problem builder's handling of checks and Xs, etc, but that's not a requirement.

smarnach committed:
Make tests pass on Travis.

smarnach commented:
@dseaton This XBlock conversion of ActiveTable has all its specific features now.  I haven't reimplemented all features that are provided by the JSInput module itself.  We don't get these features for free any more.  These features include
1. Setting a maximum number of attempts (and show a "Save" button when a maximum is set)
2. Optionally show a "Reset" button
3. Timer between attempts to force students to wait a certain amount of time before retrying
4. Randomization of problem content
5. "Show Answer" button (with eight different configuration settings)

Number one would take around 2 hours to implement.  Number 2 and 3 would be little effort on top of that, and the last two options would probably take more than two hours each (hard to estimate exactly).

What subset of these features would you like us to reimplement on the XBlock version?

cc @antoviaque

dseaton commented:
@smarnach Good point about the cost of not using JSinput. 
1. Yes.
2.  No, but would we ever need it?
3. No
4. No (for now - may need this later)
5. I think we need a sketch for the hour requirement. Honestly, we only ever use this feature in finished, or attempted mode. Does that simplify? Furthermore, would the answers appear in the table? Or would it be just showing a written solution like in any problem? The latter may be fine.

@antoviaque Let's do an hour review with Julie soon.

antoviaque commented:
@dseaton Thanks - I'll check on the number of hours spent so far and let you know.

@smarnach I'll let you reply for the hours estimates - be conservative though, especially considering that we still have to go through the edX review, which might add further work on these.

smarnach committed:
Add maximum number of attempts and Save button.

smarnach commented:
@dseaton I implemented the maximum attempt count and the save button, which took about 2.5 hours.

I looked at the "Show answer" button again, and supporting the two modes you mentioned, showing the answer as a text fragment below the table just as in the JSInput version is relatively straight-forward.  I think this will be possible in two hours including review time.

Regarding the "Reset" button – if you don't use it, you don't need it.  It should be easy to add at any point.

smarnach committed:
Validate the number of entries in column widths and row heights.

smarnach committed:
Remove Reset button for the column widths and row heights fields.

It's not working for these fields, while it works for all other fields.  This
seems to be a bug in xblock-utils.

smarnach committed:
Rename table_definition field to content for nicer XML representation.

smarnach committed:
Add workbench example scenario.

smarnach commented:
@Kelketek Regarding the non-functional reset buttons, this seems to be a bug in xblock-utils.  I currently don't have time to track it down, so I simply disabled the buttons for the two affected fields.  (It works on all other fields.)

dseaton commented:
@smarnach Thanks for the updates. Keep me posted on progression.

smarnach committed:
Fix order of indices in cell ids to be consistent with the old version.

smarnach committed:
Fix validate_field_data unit test.

smarnach committed:
Add first integration tests.

smarnach commented:
@Kelketek This is ready for another review.  I didn't mess around with the commit history, so you can restrict the review to the commits added since the last time you took a look.

Kelketek commented:
@smarnach The latest changes make sense to me, but the Travis builds still aren't working.

smarnach committed:
Update test requirements to make integrations tests pass on Travis.

smarnach commented:
@Kelketek Travis is passing now.  Could you please take another look?

sarina commented:
@smarnach We have the ospr at https://openedx.atlassian.net/browse/OSPR-923 - you've made 3 copies of ospr tickets, please just use ospr-923.

smarnach commented:
@sarina Sorry about that.  I just made yet another one, manually.  The Heroku bot is broken; it didn't give me any ticket number, and didn't update the PR with the ticket number, and only served 500s instead.  It's been doing that for a while, but so far it at least updated the PR with the ticket number in case it was successful creating one (and then served a 500 anyway).

sarina commented:
@smarnach You can only use the heroku page to make tickets for repos that the bot has permission to access and have it work smoothly; the bot can only post to repos we whitelist for it (we don't want something happening and the bot's account to get suspended). In the case of a repo that isn't under the edx/ org, you need to just go to JIRA and make the ticket manually. Honestly, please quit creating tickets. OSPR-923 has been made (as well as 922, 924, 925 and now 926). We are using 923.

smarnach commented:
@sarina I certainly didn't do this on purpose.  I've got some 500s from the bot in the past without tickets being created, and retrying helped in those cases.  I just couldn't tell that this time I was getting a different kind of 500 – from my persepective it looked identical.  It might be worth adding helpful status messages to the bot to avoid problems like this in the future.

smarnach commented:
@sarina I just noticed you misunderstood my previous comment – I had created the ticket manually _before_ I saw your comment here, and I marked it as Rejected myself immediately after that.  I didn't create yet another ticket in response to your comment, in case that is the impression you got. :)

sarina commented:
Good point. The bot wasn't designed with any non-edx repo use cases in mind. And the 500 pages could certainly use some work. At minimum, I'll try to put in some language to indicate that the bot shouldn't be used to make tickets on non edx repos.

Kelketek commented:
:+1:

sarina commented:
@smarnach : I'm working with Leslie looking at this, and we'd like to see what Davidson is using this for or wants to do with this.

antoviaque commented:
@sarina Replied in the email thread, but let me know if there is anything specific you'd like to know. We can also pass on questions as needed.

sarina commented:
Thanks for the email Xavier. I've passed it on to Leslie. My attention is
spread really thin currently across about 25 bodies of work, so I am
focused primarily on the vector drawing xblock as I think the a11y
concerns, and work that will need to be done, will be larger.

On Fri, Nov 6, 2015 at 3:42 AM, Xavier Antoviaque notifications@github.com
wrote:

> @sarina https://github.com/sarina Replied in the email thread, but let
> me know if there is anything specific you'd like to know. We can also pass
> on questions as needed.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/open-craft/xblock-activetable/pull/1#issuecomment-154345406
> .

smarnach commented:
@sarina I added a paragraph to the PR description answering the questions you asked by email:

> The XBlock is implemented in a straight-forward way. All data is stored in XBlock fields. The content is definied by a table definition in a Python-like syntax and some supporting fields. Student state consists of the answers students gave, the correctness of these answers and the score awarded. The amount of data (both content and student state) is very small (likely to be far less than 1K per problem per student). Since the amount of data is so small, and the XBlock doesn't do anything that is computationally intensive, we don't expect that there are any performance bottlenecks. We haven't performed any load testing.

Is there anything else needed for the review?

sarina commented:
Not at this time. I'm busy travelling this week, and I'm pushing on Mark to
finish his a11y review of 2 other xblocks so he can focus on reviewing this
and the vector draw block hopefully next week. Django 1.8 should land next
week, and then I'll be able to lean on Ned to help look at the xblock
itself. Thanks for the additional info.

On Tue, Nov 10, 2015 at 11:31 AM, Sven Marnach notifications@github.com
wrote:

> @sarina https://github.com/sarina I added a paragraph to the PR
> description answering the questions you asked by email:
> 
> The XBlock is implemented in a straight-forward way. All data is stored in
> XBlock fields. The content is definied by a table definition in a
> Python-like syntax and some supporting fields. Student state consists of
> the answers students gave, the correctness of these answers and the score
> awarded. The amount of data (both content and student state) is very small
> (likely to be far less than 1K per problem per student). Since the amount
> of data is so small, and the XBlock doesn't do anything that is
> computationally intensive, we don't expect that there are any performance
> bottlenecks. We haven't performed any load testing.
> 
> Is there anything else needed for the review?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/open-craft/xblock-activetable/pull/1#issuecomment-155477807
> .

sarina commented:
@smarnach - @cptvitamin is trying to do an a11y review and your sandbox is down. Could you bring it back up and then ping him?

smarnach commented:
@cptvitamin I fixed the sandbox, please take another look.

cptvitamin commented on activetable/templates/html/activetable.html:
```diff
@@ -0,0 +1,47 @@
+    </colgroup>
+    <thead>
+      <tr style="height: {{ head_height }}px;">
+        {% for cell in thead %}<th>{{ cell }}</th>{% endfor %}
```
`<th>` elements should include a scope attribute.  In this case `scope="col"`

cptvitamin commented on activetable/templates/html/activetable.html:
```diff
@@ -0,0 +1,47 @@
+          {% if cell.is_static %}
+          {{ cell.value }}
+          {% else %}
+          <input type="text" style="height: {{ cell.height }}px;" size=1
```
All form fields MUST have an accessible `<label>`.  In the example I saw, the cell in the first column would be an appropriate label.  However, I don't know if that will always be the case.

smarnach replied:
@cptvitamin No, in general the column headers will make a far better label than the first column, so I will use them instead.

smarnach replied:
@cptvitamin Is it acceptable to use an `aria-label` attribute on the input instead of a `label` tag?  I don't want to show the label to sighted users, so using `aria-label` seems like the easiest option.

cptvitamin replied:
@smarnach aria-label is theoretically an acceptable approach.  However, real world implementation support varies.  It especially gets questionable in the mobile context.  I would prefer the standard `<label>` technique.  You can hide the label offscreen with `class="sr"` .  If that is not an option, I would be OK with going with an aria-label solution.  Unfortunately, many automated testing tools will mark it as a failure.

smarnach replied:
@cptvitamin Fine, if a `label` tag is better, I'll go with that.  It works fine with the `sr` class.

I hesitated a bit since the only way to link the label and the input is using an id for the input, so I'm aggravating the problem with multiple ActiveTables on one page not being supported.  If we want to support multiple tables, we'll have to add some hash to all ids anyway, so this really won't be an additional issue.  (I can't include the input in the label because of the `sr` class.)

cptvitamin commented on activetable/templates/html/activetable.html:
```diff
@@ -0,0 +1,47 @@
+  {% else %}
+  <p>This component isn't configured properly and can't be displayed.</p>
+  {% endif %}
+  <p class="status"></p>
```
This `<p>` should have an `aria-live="polite"` attribute.

cptvitamin commented on activetable/templates/html/activetable.html:
```diff
@@ -0,0 +1,47 @@
+  <p>This component isn't configured properly and can't be displayed.</p>
+  {% endif %}
+  <p class="status"></p>
+  <div class="status-message"></div>
```
This `<div>` should have an `aria-live="polite"` attribute.

cptvitamin commented on activetable/templates/html/activetable.html:
```diff
@@ -0,0 +1,47 @@
+    <button class="check"><span class="check-label">Check</span><span class="sr"> your answer</span></button>
+    {% if max_attempts %}
+    <button class="save">Save<span class="sr"> your answer</span></button>
+    <div class="submission-feedback"></div>
```
This `<div>`  should have an `aria-live="polite"` attribute.

cptvitamin commented on activetable/templates/html/activetable.html:
```diff
@@ -0,0 +1,47 @@
+<div class="activetable_block">
+  {% if help_text %}
+  <div class="activetable-help" style="width: {{ total_width }}px;">
+    <button id="activetable-help-button">+help</button>
```
The + and - are not ready by screen readers reliably.  This button should include an `aria-expanded` attribute that informs the user of the current state of the disclosure (help text).  There should also be an explicit reference to the associated `<p>` element with an `aria-controls` attribute and an `aria-haspopup="true"` attribute.

These IDs seem to be hard coded.  What happens if there are multiple active table problems on a single page?  IDs have to be unique, especially if they are being referenced by other elements, otherwise the whole reference system breaks down.

smarnach replied:
@cptvitamin I added the `aria-…` properties you mentioned.

And you are right, we currently don't support multiple tables per page.  Some of the ids could easily be turned into classes.  Most lookups are performed relative to the top-level XBlock div anyway, so we wouldn't get conflicts with other tables on the same page.  However, the `p` element with the help text needs to be referenced by id in the `aria-controls` label, so turning it into a class is not an option.  I guess I will have to add some hashes to the few ids we need, and pass them around everywhere.

smarnach commented:
@cptvitamin Thanks for the review and your comments.  I'll address them today.

smarnach commented:
@sarina We'd like to get this merged by mid-December, which is by the end of next week.  Would it be possible to schedule a timely review, preferably this week, so I would still have time to address any comments?

nedbat commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+django>=1.4, <1.5
```
We're on 1.8 now

nedbat commented on activetable/activetable.py:
```diff
@@ -0,0 +1,297 @@
+    )
+    default_tolerance = Float(
+        display_name='Default tolerance',
+        help='The tolerance in pecent that is used for numerical response cells you did not '
```
"percent"

nedbat commented on activetable/activetable.py:
```diff
@@ -0,0 +1,297 @@
+    column_widths = String(
+        display_name='Column widths',
+        help='Set the width of the columns in pixels.  The value should be a Python-like list of '
+        'numerical values.  The total width of the table should not be more than 800. No value '
```
Instead of "No value", say, "Omitting this value"

nedbat commented on activetable/activetable.py:
```diff
@@ -0,0 +1,297 @@
+        self.response_cells = {}
+        for row, height in zip(self.tbody, self._row_heights[1:]):
+            row['height'] = height
+            if row['index'] & 1:
```
This crossed my eyes for a moment.  Maybe "% 2" instead?

smarnach replied:
I don't care, so I can go with `% 2` as well.  For me, the intuitive way of thinking about this was "use the last bit", but this isn't the first time that other people's intuition is different. :)

nedbat commented on activetable/activetable.py:
```diff
@@ -0,0 +1,297 @@
+        'max_attempts',
+    ]
+
+    # Dictionary mapping cell ids to the student answers.
```
Are cell IDs stable while editing? Inserting a row in the table will make students' existing answers mismatched, right?  Is this understood by the course team? Can we say something in the help text?

smarnach replied:
Yes, cell IDs are really just strings like `cell_i_j` where `i` is the row index and `j` is the column index, so editing the table structure will invalidate student data.  I think this is well understood by the course team, but it doesn't hurt to add a comment to the help text.

nedbat commented on activetable/activetable.py:
```diff
@@ -0,0 +1,297 @@
+
+    def student_view(self, context=None):
+        """Render the table."""
+        self.parse_fields()
```
I'm surprised to see the fields being parsed each time.  Not sure why that bothers me, but it seems like significant work that could be done just once.

smarnach replied:
The result of parsing is a hierachical structure of Python data types that are not JSON-serializable, so I don't see an easy way of doing the parsing only once.  I initially added the parsing step to `__init__()`, but found out that [XBlocks are not supposed to do that](http://edx.readthedocs.org/projects/xblock-tutorial/en/latest/concepts/fields.html#initializing-fields), and adding an `__init__()` method indeed broke things in weird ways.

nedbat commented on activetable/activetable.py:
```diff
@@ -0,0 +1,297 @@
+            except ParseError as exc:
+                add_error('Problem with column widths: ' + exc.message)
+            else:
+                if thead is not None and len(column_widths) != len(thead):
```
Did you want to check the 800px limit?

smarnach replied:
I intentionally didn't check it.  The 800px limit is a rule of thumb that works well when using this XBlock on edx-platform on desktop, but people might have other requirements anyway, so it's really up to them to decide whether it looks good enough for them.

nedbat commented on activetable/cells.py:
```diff
@@ -0,0 +1,77 @@
+class NumericCell(Cell):
+    """A numeric response cell."""
+
+    is_static = False
```
It'd be nicer to move this to Cell, and let StaticCell be the only one that needs to override it.

nedbat commented on activetable/cells.py:
```diff
@@ -0,0 +1,77 @@
+
+    def check_response(self, student_response):
+        """Return a Boolean value indicating whether the student response is correct."""
+        return student_response == self.answer
```
Is the course team really OK with strict equality here? No .strip() or .lower() ?

smarnach replied:
My suspicion is that this feature hasn't received a lot of use in the JSInput version of this code (which also used exact string matching).  It would be nice to use some configurable fuzzy matching here, but that's probably for a later version of the code.  For now, I'll simply add `.strip()`, since `.lower()` might not be desired in all contexts.

nedbat commented on activetable/parsers.py:
```diff
@@ -0,0 +1,102 @@
+    except SyntaxError as exc:
+        raise ParseError(exc.msg)
+
+    row_iter = iter(_ensure_type(expr.body, ast.List).elts)
```
I like the code in this function :)

nedbat commented on README.md:
```diff
@@ -0,0 +1,57 @@
+well.  Significant digits are counted started from the first non-zero digit specified by the
+student, and include trailing zeros.
+
+    String(answer='<correct answer>')
```
"String" is pretty geeky, wouldn't course teams be happier with "Text"?

smarnach replied:
@dseaton Are you fine with renaming `String` to `Text`, or would that cause any migration problems for you?

nedbat commented on activetable/parsers.py:
```diff
@@ -0,0 +1,102 @@
+        raise ParseError('invalid cell input type: {}'.format(cell_type))
+    try:
+        return cell_class(**kwargs)
+    except TypeError as exc:
```
Is TypeError the only possible failure here?  This would be a good place to use "except Exception".

smarnach replied:
The tests above are quite strict, so if I made no mistake, the only error state left is misnamed keyword parameters, which would result in a `TypeError`.  But you are right, it might be safer to catch `Exception` here, so I'll do that.

nedbat commented on activetable/parsers.py:
```diff
@@ -0,0 +1,102 @@
+    This is used to parse the column_widths and row_heights lists entered by the user.
+    """
+    try:
+        lst = ast.literal_eval(source)
```
literal_eval trivia: for some reason, in Python 3.5, this works: `ast.literal_eval("[1+1]")` to produce [2], but this fails: `ast.literal_eval("[1*1]")`.  No idea why.

smarnach replied:
This is related to the fact that `literal_eval()` supports complex numbers, so it needs at least limited support for the binary `+` operator.

The difference between Python 2.x and Python 3.x is due to a change in the parser.  Python 2.x parses negative integers as single nodes in the AST (which is the only way to get corner cases for the distinction between `int` and `long` right).  In Python 3.x, this disctinction is no longer needed, so they removed the negative integer literals, and instead parse them into two nodes in the AST now, one for the unary minus operator and one for the integer literal.  This new parsing behaviour required changes to the implementation of `literal_eval()`, since now explicit support for the unary minus operator was needed, and when they changed this, they also relaxed the rules for what operands for the binary plus are allowed a bit.

Here's the patch: https://hg.python.org/cpython/rev/884c71cd8dc6/

nedbat replied:
Wow, thanks for the detail! :)

nedbat commented on activetable/static/js/src/activetable.js:
```diff
@@ -0,0 +1,96 @@
+            $status.text('incorrect');
+            $status_message.text(
+                'You have ' + data.num_correct_answers + ' out of ' + data.num_total_answers +
+                ' cells correct.'
```
We still have no answer for how to localize XBlocks, right?

smarnach replied:
The current recommendation seems to be to mark string in Python files, but not anywhere else.

nedbat commented on run_tests.py:
```diff
@@ -0,0 +1,52 @@
+"""Run tests for the ActiveTableXBlock.
+
+This script is required to run the Selenium tests inside the xblock-sdk workbench
+because the workbench SDK's settings file is not inside any python module.
```
I would like to understand this better, can you email me with an explanation so we can look into fixing it in the workbench?

smarnach replied:
I'll have to look into this a bit myself.  I tried briefly to get the Selenium tests running without this script, and when that failed, I simply copied the script from our Problem Builder XBlock (which I didn't write), which just worked, so I didn't give it too much thought.

nedbat commented on tests/unit/test_cells.py:
```diff
@@ -0,0 +1,35 @@
+    def test_string_cell(self):
+        cell = StringCell('OpenCraft')
+        self.assertTrue(cell.check_response('OpenCraft'))
+        self.assertFalse(cell.check_response('giraffe'))
```
Might want to add tests with differing case, whitespace, and some non-ascii characters.

nedbat commented:
This looks good.  I wonder about parsing the table on every action, and whether the UI is right for the course team.

smarnach committed:
Adjust requirements for Django 1.8.

smarnach committed:
Address Ned's review notes.

smarnach committed:
Rename "String" cells to "Text" cells.

smarnach committed:
Address Mark's review notes.

smarnach commented:
@cptvitamin I addressed all your comments.  The new version is live on the sandbox (linked in the PR description).  Could you please take another look?  We'd like to get this merged by the end of the week, so I'd appreciate if you could look into this soon.

smarnach commented:
@nedbat I think I've addressed your comments as well.  The course team explicitly asked for the UI to be the way it is now; they specifically didn't want a more graphical interface.  Could you please take another look?

nedbat commented:
:+1:

nedbat commented:
(To clear, my thumbs-up means that I am OK with you merging this code.  I believe you have the authority.)

smarnach commented:
@cptvitamin Could you please take another look at this PR?  We'd like to get this into Dogwood, so it would be ideal if I could merge it today.
